Apply redesign to header

stevenwilkin opened the pull request:
A reworking of [~stevenwilkin/webdm/new-header/+merge/295574](https://code.launchpad.net/~stevenwilkin/webdm/new-header/+merge/295574) taking @bartaz 's suggestions in consideration.

All occurrences of `!important` are now gone and all overriding of Vanilla's styles has been confined to `lib/vanilla-overrides.scss`.

Considering most of the CSS declarations are found in the overrides file does make me consider the suitability of using the Vanilla header as a base for the new design. Any opinions @bartaz ?

# larger viewport

![image](https://cloud.githubusercontent.com/assets/45121/16015557/7143b6b0-31c0-11e6-87b7-ed60f661c05e.png)

# smaller viewport
![image](https://cloud.githubusercontent.com/assets/45121/16015572/80f0de62-31c0-11e6-8155-29b8b93090b5.png)

bartaz commented on www/src/css/lib/vanilla-overrides.scss:
```diff
@@ -5,4 +5,50 @@
+  background-color:transparent;
+}
+
+@media screen and (max-width: 640px) {
```
As `$navigation-threshold` is set, I would suggest using it instead of hardcoding 640px.

Also vanilla source uses mostly mobile fist approach (defining mobile styles without media query and using min-width media query for desktop styles), so it may be worth adapting to same structure, but that's just minor suggestion.

stevenwilkin replied:
Good catch on using the SCSS variable here.

I hadn't considered the mobile-first approach to be honest but now that you point it out it makes sense. Subtle but logical. I'll update things accordingly as it would mean having fewer media queries and should be easier to reason about.

stevenwilkin replied:
So `lib/vanilla-includes.scss` isn't imported into `lib/vanilla-overrides.scss`, the 2 files are merely concatenated by gulp and so the variable isn't available. Do we want keep this structure?

bartaz replied:
@stevenwilkin You are right. I guess it would make sense to import vanilla-includes here to make the most of variables, etc.

Or maybe it would make more sense to import `vanilla-overrides` into `vanilla-includes` - so that vanilla-includes would be one file to wrap all things vanilla.

stevenwilkin replied:
Having `vanilla-includes` as the single hook into the Vanilla world which then pulls in the overrides seems to do the job nicely.

bartaz commented on www/src/css/lib/vanilla-overrides.scss:
```diff
@@ -5,4 +5,50 @@
+  }
+
+  .banner .logo {
+    margin:0 0 0 20px;
```
`.banner .logo` margin styles - here and above (20px or 10px) seem not very different from what vanilla has by default. Do they need to be overridden at all? Won't defailt vanilla margins be good enough?

stevenwilkin replied:
Seems like a reasonable approach to take, any opinions @kwm14 @jamiedawsonyoung ?

jamiedawsonyoung replied:
If the margin values are very close (but slightly different) they were almost certainly intended to be Vanilla values. If in doubt, please stick to Vanilla values :-) @kwm14 Does that sounds about right?

kwm14 replied:
Yeah that sounds right to me @jamiedawsonyoung. The values are close so @stevenwilkin if we can use as much styling from Vanilla that would be great! Thanks

stevenwilkin replied:
Grand, I'll let Vanilla dictate the margins here. We'll find out down the line what the effect will be on the snap grid etc.

bartaz commented on www/src/css/lib/vanilla-overrides.scss:
```diff
@@ -5,4 +5,50 @@
+  background-color:transparent;
+  font-size:initial;
+  border-left:0;
+  padding:14px 12px 8px 12px;
```
Do we need to override `font-size` and `padding`? Won't default vanilla values be good enough?

I know that on designs they are bit different than in vanilla - question is - is it worth overridding them with a cost of code quality (more overrides) and also being not so consistent with vanilla.

stevenwilkin replied:
These were set to achieve a specific size and position in relation to the logo for the most part.

Without these overrides the nav looks roughly like this:

![image](https://cloud.githubusercontent.com/assets/45121/16077782/232c755a-3325-11e6-9ed2-f4d9a1c2f411.png)

Versus with the overrides:

![image](https://cloud.githubusercontent.com/assets/45121/16077793/2e26fea8-3325-11e6-9a0a-80974e6ea3b4.png)

Maybe @jamiedawsonyoung or @kwm14 can chip in here?

kwm14 replied:
@stevenwilkin they're close in terms of visual appearance, but looking closer the override (Vanilla version) fits more inline and would be consistent across all Cloud products. @jamiedawsonyoung do you agree?

jamiedawsonyoung replied:
Tdlr: without the overrides please. Always err on the side of Vanilla styles.

The only overrides we really want (not landing with this) is the addition of the icons next to the menu labels which 'I think' are still in the designs. To answer this question though, if it's this close to Vanilla it's almost 100% likely we were trying to use Vanilla styles but didn't get it right.

If we're not happy (in the design team) with margins / padding etc then we should change Vanilla and have the changes propagate to the different products.

@kwm14 I can catch you up on this when you're back… 

bartaz commented on www/src/css/lib/vanilla-overrides.scss:
```diff
@@ -5,4 +5,50 @@
+
+.banner .nav-primary ul li a:hover,
+.banner .nav-primary ul li a:link:hover,
+.banner .nav-primary ul li a:visited:hover {
```
That's quite strange to use both pseudo classes at the same time `:link:hover` `:visited:hover`. Isn't just `:hover` enough? Or does structure of vanilla styles require for such double overrides?

stevenwilkin replied:
This was me copy'n'pasting the selector from the Chrome web inspector, I'll give it a go without all the pseudo-classes

stevenwilkin replied:
Seems the pseudo-classes are actually needed to deal with the selectors generated from the Vanilla SCSS

bartaz replied:
@stevenwilkin That's a pity. Yes, I see now that this is exactly how vanilla generated selectors look like (which to me looks like a bug). But in such case it has to stay like this.

Maybe it would be worth adding a short one line comment there to explain that these selectors are based on vanilla and are needed for override. So the next person looking there won't have to think about it.

stevenwilkin replied:
Good call on the comment

bartaz commented on www/src/css/banner.css:
```diff
@@ -1,121 +1,37 @@
-  right:0;
-  height:2px;
-  transition: opacity 0.3s;
+.banner .nav-primary .b-banner__nav-item_active {
```
That looks like another vanilla override (based on `.banner .nav-primary`) - shouldn't it be in vanilla-overrides.scss?
I understand that that would probably mean changing BEM `.b-banner_nav-item_active` to `active` class used by vanilla.

stevenwilkin replied:
You're right, I tried to keep the original class in the template but changing it would be tidier

bartaz commented on www/src/css/banner.css:
```diff
@@ -1,121 +1,37 @@
 
-.b-banner__nav-item_active::before {
-  opacity:1;
+.b-banner__nav-item svg {
```
To make these BEM styles cleaner it would be worth avoiding element names in selectors and add classes to them too (for example `b-banner__nav-item-icon instead` of `svg`). Same for `a` and `span` below (`nav-item-link`, `nav-item-label` maybe)?

stevenwilkin replied:
I'l give it a go!

bartaz commented:
I added couple of question/thoughts inline in code comments.

The split of vanilla overrides and rest of styles looks good.

I agree that looking at amount of overrides we have for navigation styles it's worth considering usage of vanilla header styles.
I don't know how much of vanilla header styles do we actually use. Seems like the biggest feature of vanilla header (responsive navigation turning into dropdown) is not used here, so we need to override that with custom responsive styles (showing icons on mobile). Also nav item styles are bit different (here we use white background, darker on hover, border on active items). So it seems like apart from overall looks that are similar to vanilla (but still bit different in terms of margins/paddings/font-size) snapweb navigation is quite different from vanilla nav.

I would suggest checking how much code would it require to remove vf-header and writing new header styles (based on vanilla). If we need less or similar amount of code to write header styles from scratch that we have in overrides it would be good to not use vf-header. If overrides are significantly less code, I would stick to using vanilla header.

stevenwilkin commented:
I've pushed up some fixup commits to address the points discussed so far. They will be squashed before merging.

Looking at Vanilla's [modules/_header.scss](https://github.com/ubuntudesign/vanilla-framework/blob/master/scss/modules/_header.scss) the `.banner` declaration totals >160 lines whereas our `banner.css` and overrides files have roughly half that at the moment.

I unscientifically guess that re-implementing the header without `vf-header` could end up with more lines of CSS in the project than proposed in this PR. I dislike all the overrides but it may be easier to live with them especially when it comes to upgrading to future versions of Vanilla. Other opinions are more than welcome though.

barrymcgee commented:
I agree that sticking with the overrides is problem the best course of action in the short-term given that Vanilla header pattern will change significantly in coming updates.

stevenwilkin commented:
Rebased on master. Just [one outstanding question](https://github.com/snapcore/snapweb/pull/9#discussion_r67156234) for @jamiedawsonyoung

stevenwilkin commented:
I believe I've addressed all the feedback and reduced this PR down to the minimum amount of work on top off what's provided by Vanilla.


# viewport >= 640px
![image](https://cloud.githubusercontent.com/assets/45121/16188844/bf354e60-3701-11e6-92f2-1b7b2bfcb385.png)

# viewport < 640px
![image](https://cloud.githubusercontent.com/assets/45121/16188854/cd954122-3701-11e6-8a83-68af9f87cd27.png)

bartaz commented:
@stevenwilkin Nice work! Good to see `vanilla-overrides` shrinking down to reasonable chunk.

stevenwilkin commented:
Great, we should be good to merge then if @jamiedawsonyoung or @kwm14 gives us the nod.

kwm14 commented:
@stevenwilkin sat down with @jamiedawsonyoung and we reviewed redesign of the header, it looks good so happy for you to merge.

I know your working towards updating code to Vanilla before moving onto new design updates. But I just wanted to flag up about an nav update within the new design files I sent through last week, we now have icons next to heading labels in the nav above 640px viewport.

stevenwilkin commented:
Good stuff @kwm14.

The priority is to ship the design as it currently stands in the prototype within in the next week so updating the nav will not be addressed until afterwards.

kwm14 commented:
@stevenwilkin Ok great. Thanks for updating me on the current prototype build.
